### DISCUSSION TITLE
[General] <How to interact with a contract on Ethereum from a contract on zkSync?>

### DISCUSSION THREAD

**Discussion by kientt265** [zkSync-Community-Hub/zksync-developers#751] [General] [answered]
### Team or Project

Hospital (transfer patient profiles through zkSync)

### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement and the issue you are running into.

I'm confused trying to interact between ethereum and zksync. Please give me detailed steps to interact between these two layers. For example: I have a doctor object created in a contract on ethereum and I want that doctor object to be able to perform the functions I have pre-set up on zksync just for that doctor object. Please tell me what I should do

### Repo Link (Optional)

https://github.com/kientt265/-web3-project-zk

**Comment by alhenaai** [comment]
To interact between a contract on Ethereum (L1) and a contract on ZKsync (L2), you can follow these steps:

1. **Deploy Your Contracts**: Ensure you have your contracts deployed both on Ethereum and ZKsync. Your Ethereum contract should have the necessary functions to interact with the ZKsync contract.

2. **Use ZKsync-Ethers**: Utilize the `zksync-ethers` library, which provides helper methods to facilitate L1 to L2 interactions. This library will help you send transactions from Ethereum to ZKsync.

3. **Set Up Communication**: Implement a function in your Ethereum contract that calls the ZKsync contract. You will need the ABI and the address of the ZKsync contract. Use the `zksync-ethers` to create a transaction that targets your ZKsync contract.

4. **Call Functions**: From your Ethereum contract, call the function intended to interact with the ZKsync contract. This might involve sending specific data or calling specific methods predefined in the ZKsync contract.

5. **Handle Responses**: Ensure your ZKsync contract can handle incoming calls and data from the Ethereum contract and can perform necessary actions or state changes.

For a practical example and more detailed guidance, you can refer to the tutorial on [how to send a transaction from L1 to L2](https://code.zksync.io/tutorials/how-to-send-l1-l2-transaction), which includes common use cases and step-by-step instructions.

For further development discussions or specific issues, consider visiting the [ZKsync developers discussions on Github](https://github.com/zkSync-Community-Hub/zksync-developers/discussions).

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by cpb8010** [comment, marked as answer]
There's no direct path to call from your custom L1 contract (on Ethereum) into or from L2 (on ZKsync Era). You'll need to communicate via transactions via another contract (the specifics of which will depend on your application).

The path you might want to take if you want to update these two contracts is to have your dapp read from your Ethereum contract and send transactions to ZKsync to interact with the contract there. Making ZKsync transactions directly is going to be cheaper and faster, but if you only want to perform L1 transactions, you can have your L1 contract call [requestL2Transaction](https://github.com/code-423n4/2023-10-zksync/blob/ef99273a8fdb19f5912ca38ba46d6bd02071363d/code/contracts/ethereum/contracts/zksync/facets/Mailbox.sol#L236). There's a great tutorial here: https://code.zksync.io/tutorials/how-to-send-l1-l2-transaction that walks you through the exact steps.
